contrib/babelfishpg_tsql/src/pl_exec.c

```diff
@@ -4597,10 +4597,79 @@ exec_stmt_execsql(PLtsql_execstate *estate,
 	bool		enable_txn_in_triggers = !pltsql_disable_txn_in_triggers;
     StringInfoData query;
 	Oid			current_user_id = GetUserId();
+	PLExecStateCallStack *top_es_entry;
+	bool		need_path_reset = false;
+	char		*cur_dbname = get_cur_db_name();
+	/* fetch current search_path */
+	List 		*path_oids = fetch_search_path(false);
+	char 		*old_search_path = flatten_search_path(path_oids);
+	char 		*new_search_path;
+	char 		*physical_schema;
+	char		*dbo_schema;
 
 	if (stmt->is_cross_db)
 		SetCurrentRoleId(GetSessionUserId(), false);
 
+ 	top_es_entry = exec_state_call_stack->next;
+	if (stmt->is_dml)
+	{
+		while(top_es_entry != NULL)
+		{
+			/* traverse through the estate stack. If the occurrence of
+			 * exec in the call stack, update the search path.
+			 */
+			if(top_es_entry->estate && top_es_entry->estate->err_stmt &&
```

Comment:
why are we doing it this way rather than setting right search path when a procedure starts execution?

Reply:
If we set the search path at the start of the execution, it will look for the procedure name based on that search path which is not required. Also, we need to set the search path differently based on the DDL/DML statements being referred inside the procedure.

Reply:
I do not understand the diff req for DDL and DML, what if we add both physical and DBO schema to search paths, will it create an issue for DDL?

Reply:
If we do that for DDL(CREATE) and we have that inside sch1.p1, we would end up creating the table in sch1 but not default [which is not the required behaviour]. In case of DML(SELECT), we need to select from sch1.t1 and if not found we need to search for dbo.t1 but not default.  This is why we need separate handling for DDL and DML.

Reply:
OK, I would still suggest to put this logic in a separate function with single loop and have conditions inside to set the correct search path, rather than repeating the same loop twice.

Reply:
Moved this logic to a function and used a single loop.